In [5]:
from typing import Tuple, Iterable, List, Dict, Callable, Any, Optional, Union
import random

# <center>Функции</center>

# 1. Метрики

Данное задание призвано как проверить навыки работы с функциями, так и ознакомить Вас с метриками, которые используются во время тренировки моделей Supervised Learning.

### 1.1 Gini score

Критерий Gini показывает, насколько однородным является выбранная группа, и расчитывается по следующей формуле:

$$Gini = 1 - \sum p_{j}^{2}$$

где $p_{j}$ - это частота каждого класса в данной группе (или, иными словами, вероятность того, что случайно взятый экземпляр из этой группы принадлежит к данному классу).  
  
Например, у нас есть группа из 20 шариков, 10 из которых - синие, 7 - зелёные, и 3 - красные. В данном случае критерий Gini будет равен:

$$Gini = 1 - ((\frac{10}{20})^{2} + (\frac{7}{20})^{2} + (\frac{3}{20})^{2}) = 1 - 0.395 = 0.605$$

Критерий Gini может принимать значения от 0 (абсолютно гомогенная группа, состоящая из одного класса) до 1 (максимальная неопределённость, множество классов с одинаковыми вероятностями выпадения)

#### Задание: написать функцию, которая получает на вход список целых чисел, и вычисляет значение критерия Gini

In [2]:

def get_periodicity(items: Iterable[Union[str, float]]) -> Iterable[Union[str, float]]:  

    items_list = list(items)
    items_set = set(items)
    
    periodicity_items = [
        items_list.count(item)
        
        for item 
        in items_set
    ]
    
    return periodicity_items


def gini_score(periodicity_items: Iterable[int]) -> float:
    items_list = list(periodicity_items)
    sum_items = sum(items_list)
    
    gini_score_value = sum(list(map(lambda x: pow(x / sum_items, 2), items_list)))
    
    return 1 - round(gini_score_value, 3)

    
def count_gini_score(items: Iterable[Union[str, float]]) ->  float:  
    return round(gini_score(get_periodicity(items)), 3)

    
print(count_gini_score([1, 2, 1, 2, 0, 1]))
print(count_gini_score(["A", "A", "A", "A", "B", "A", "A", "C", "D"]))
print(count_gini_score(["С", "С", "С", "З", "К", "С", "С", "С", "С", "С", "З", "З", "З", "З", "З", "З", "С", "С", "К", "К"]))


0.611
0.519
0.605


---

### 1.2 Entropy

Энтропия измеряет степень неопределённости группы, и, конкретно в нашем случае, описывает количество информации, требуемое для описания распределения (distributon) в группе.  
Формула (средней) энтропии:

$$Entropy = - \sum p_{j}\log p_{j}$$

как и в случае с критерием Gini, $p_{j}$ - это частота каждого класса в данной группе.  
  
В том же примере с шариками (группа из 20 шариков, 10 из которых - синие, 7 - зелёные, и 3 - красные), уровень энтропии будет равен:

$$Entropy = -\ (\frac{10}{20} * \log(\frac{10}{20}) + \frac{7}{20} * \log(\frac{7}{20}) + \frac{3}{20} * \log(\frac{3}{20})) = - \ (-0.347 - 0.367 -0.285) = 0.999$$

Энтропия может принимать значения от 0 (абсолютно гомогенная группа, состоящая из одного класса) до бесконечности (максимальная неопределённость), хотя на практике максимальное значение ограничено логарифмической функцией и достигает величины в 20-30 единиц (можете попробовать запустить её со списком из миллиона уникальных значений).

#### Задание: написать функцию, которая получает на вход список целых чисел, и вычисляет значение Энтропии

In [36]:
import math

def entropy_score(periodicity_items: Iterable[int]) -> float:
    items_list = list(periodicity_items)
    sum_items = sum(items_list)
    
    entropy_score_value = sum(list(map(lambda x: (x / sum_items) * math.log(x / sum_items), items_list)))
    
    return (-1) * round(entropy_score_value, 3)

def count_entropy_score(items: Iterable[Union[str, float]]) ->  float: 
    return entropy_score(get_periodicity(items))

   
print(count_entropy_score([1, 2, 1, 2, 0, 1]))
print(count_entropy_score(range(5_000)))
print(count_entropy_score(["A", "A", "A", "A", "B", "A", "A", "C", "D"]))
print(count_entropy_score(["С", "С", "С", "С", "С", "С", "С", "С", "С", "С", "З", "З", "З", "З", "З", "З", "З", "К", "К", "К"]))

1.011
8.517
1.003
0.999


---

### 1.3 Mean Squared Error (MSE)

Среднеквадратичная ошибка (Mean Squared Error, MSE) - это метрика, используемая в задачах регрессии, для оценки разницы между прогнозируемыми и фактическими значениями.  
  
MSE измеряет среднеквадратичное отклонение прогнозируемых значений от фактических значений и вычисляется путем суммирования квадратов разностей между каждым прогнозируемым и фактическим значением, а затем деления на количество наблюдений:

$$MSE = \sum_{i=0}^{n} \frac{(y^{'}_{i} - y_{i})^{2}}{n}$$

где $y^{'}_{i}$ - это предсказанное значение для $i$-того элемента, $y_{i}$ - его реальное значение, $n$ - общее количество элементов.  
  
Очевидно, что в идеальном случае, когда все предсказания в точности равны реальным значениям, **MSE** будет равен нулю (сверху же данный показатель не ограничен).

В рамках этого задания мы слегка изменим эту метрику, и вместо $y^{'}_{i}$ (отдельного предсказания для каждого элемента) будем испльзовать среднее значение всех элементов в группе - $\hat{y}$:

$$MSE = \sum_{i=0}^{n} \frac{(\hat{y} - y_{i})^{2}}{n}$$

Пример: дана группа чисел $[-0.1, -2.0 ,  1.0 ,  1.8,  0.8]$.
  
Среднее значение группы:

$$\hat{y} = \frac{-0.1-2.0+1.0+1.8+0.8}{5} = 0.3$$

Значит, среднеквадратичное отклонение будет равно:

$$MSE = \frac{(0.3 - (-0.1))^{2} + (0.3 - (-2.0))^{2} + (0.3 - 1.0)^{2} + (0.3 - 1.8)^{2} + (0.3 - 0.8)^{2}}{5} = 1.688$$

#### Задание: написать функцию, которая получает на вход список чисел, и вычисляет среднеквадратичную ошибку:

In [3]:
def mean_value(values: list) -> float:
    return sum(values) / len(values)
    

In [4]:
def mse_score(items: Iterable[Union[int, float]]) -> float:
    items_list = list(items)
    mean_items_value = mean_value(items)
    
    mse_score_value = sum(list(map(lambda x: (mean_items_value - x) ** 2, items_list)))
    
    return round(mse_score_value / len(items_list), 3)
    
mse_score([-0.1, -2.0, 1.0, 1.8, 0.8])

1.688

---

### 1.4 Root Mean Squared Error (RMSE)

Корень средней квадратичной ошибки (Root Mean Squared Error, RMSE) имеет преимущество в том, что представляет величину ошибки в единице прогнозируемого столбца, что упрощает процесс интерпретации.  
  
Если вы пытаетесь спрогнозировать сумму в долларах, корень средней квадратичной ошибки можно интерпретировать как величину ошибки в долларах.
  
Аналогично предыдущему примеру, в данном задании мы будем вместо прогнозируемых значений использовать среднее группы:

$$RMSE = \sqrt{\sum_{i=0}^{n} \frac{(y^{'}_{i} - y_{i})^{2}}{n}} \longrightarrow RMSE = \sqrt{\sum_{i=0}^{n} \frac{(\hat{y} - y_{i})^{2}}{n}}$$

#### Задание: написать функцию, которая получает на вход список чисел, и вычисляет корень среднеквадратичной ошибки:

In [22]:
def rmse_score(items: Iterable[Union[int, float]]) -> float:
    return round(mse_score(items) ** (1 / 2), 3)

rmse_score([-0.1, -2.0, 1.0, 1.8, 0.8])

1.299

---

### 1.5 Mean Absolute Percentage Error (MAPE)

Средняя процентная ошибка (МАРЕ) измеряет отклонение фактических значений от прогнозируемых значений в процентах.  
  
MAPE часто используется в бизнесе и экономике для оценки точности прогнозирования, особенно при прогнозировании спроса на товары или услуги.  
  
Преимущество данной метрики в том, что она позволяет сравнивать эффективность моделей, натренированных на данных различного масштаба.  
  
Формула для расчёта MAPE:

$$MAPE = \frac{100}{n} * \sum{\mid \frac{y^{'}_{i} - y_{i}}{y_{i}} \mid}$$

Поскольку реальное значение $y_{i}$ может быть равен нулю, то для избежания деления на ноль в некоторых случаях в знаменатель добавляется ничтожно малое число $\varepsilon$ (оно должно быть существенно ниже имеющихся реальных значений):

$$MAPE = \frac{100}{n} * \sum{\mid \frac{y^{'}_{i} - y_{i}}{y_{i}} \mid} \ \longrightarrow \ MAPE = \frac{100}{n} * \sum{ \frac{\mid y^{'}_{i} - y_{i}\mid}{\mid y_{i} \mid + \varepsilon}}$$

Другой способ избежать деление на ноль - это просто убрать из расчёта экземпляры, где реальное значение равно нулю (ибо для них применение этой метрики не имеет смысла)

Аналогично предыдущему примеру, в данном задании мы будем вместо прогнозируемых значений использовать среднее группы:

$$MAPE = \frac{100}{n} * \sum{ \frac{\mid \hat{y} - y_{i}\mid}{\mid y_{i} \mid + \varepsilon}}$$

#### Задание: написать функцию, которая получает на вход список чисел, и вычисляет MAPE:

In [104]:
def mape_score(items: Iterable[Union[int, float]]) -> float:
    items_list = list(items)
    mean_items_value = mean_value(items)
    
    items_filtered = list(filter(lambda x: x != 0, items_list))
    
    mape_score_value = sum(list(map(lambda x: abs(mean_items_value - x) / abs(x), items_filtered)))
    
    return round((100 / len(items_list)) * mape_score_value,  3)
                          
mape_score([-0.1, -2.0, 1.0, 0, 1.8, 0.8])

115.394

---

# 2. Взвешенная оценка

В некоторых ситуациях нам надо расчитать среднее взвешанное значение заданной метрики для нескольких групп. Слово *взвешенное* подразумевает, что вклад каждой группы в финальное значение пропорционален количеству элементов в ней.
  
Например: у нас есть две группы `A: [1, 1, 2, 1, 3, 2]` и `B: [3, 3, 2]`. Задача - расчитать взвешенное значение критерия Gini:

$$Gini_{total} = \frac{N_{A}}{N_{A} + N_{B}}Gini_{A} + \frac{N_{B}}{N_{A} + N_{B}}Gini_{B} = \frac{6}{9}*0.611 + \frac{3}{9}*0.444 = 0.556$$

#### Задание: написать функцию, которая получает на вход функцию метрики и несколько списоков чисел, и вычисляет взвешенное значение переданной метрики:

In [29]:
def weighted_metric_score(metric: Callable, *items: Iterable) -> float:  
    a = list(items[0])
    b = list(items[1])
    
    n_a = len(a)
    n_b = len(b)
    sum_n = n_a + n_b
    return round((n_a / sum_n) * metric(a) + (n_b / sum_n) * metric(b),3)
    
weighted_metric_score(count_gini_score, [1, 1, 2, 1, 3, 2], [3, 3, 2])

0.555

---

# 3. Разделенение группы

### 3.1 Деление группы на две

Легче всего описать это задание с вводных данных:  
  
Есть два списка: один - вспомогательный (`features, F`), состоящий из чисел, второй - основной (`targets, T`), при этом их элементы связаны попарно:

$$F_{i} \leftrightarrow T_{i}$$

Нам даётся число из вспомогательной группы `F`. Наша задача - разделить основную и вспомогательную группы так на две подгруппы так, чтобы все элементы $\color{red}{первой}$ вспомогательной подгруппы были бы меньше или равны данному числу, а все элемены $\color{blue}{второй}$ вспомогательной подгруппы - больше этого числа.
  
Например: нам данны следующие основная и вспомогательная группы:

$$
T = [1, 1, 2, 2, 1, 1, 2, 1, 3, 1, 3, 4, 1]  \\
F = [3, 1, 9, 9, 1, 5, 2, 1, 7, 1, 3, 4, 9]
$$

Допустим, из вспомогательного списка выбрано число `5`. Таким образом, эти списки будут разделены следующим образом:

$$
T = [\color{red}{1}, \color{red}{1}, \color{blue}{2}, \color{blue}{2}, \color{red}{1}, \color{red}{1}, \color{red}{2}, \color{red}{1}, \color{blue}{3}, \color{red}{1}, \color{red}{3}, \color{red}{4}, \color{blue}{1}] \rightarrow [\color{red}{1,1,1,1,2,1,1,3,4}] \ , \ [\color{blue}{2,2,3,1}] \\
F = [\color{red}{3}, \color{red}{1}, \color{blue}{9}, \color{blue}{9}, \color{red}{1}, \color{red}{5}, \color{red}{2}, \color{red}{1}, \color{blue}{7}, \color{red}{1}, \color{red}{3}, \color{red}{4}, \color{blue}{9}] \rightarrow [\color{red}{3,1,1,5,2,1,1,3,4}] \ , \ [\color{blue}{9,9,7,9}]
$$

#### Задание: написать функцию, которая получает на вход основной список `targets`, вспомогательный список `features` и "делитель" `split_value` и возвращает результат разбиения основного списка `sub_targets_1`, `sub_targets_2`:

In [30]:
def split_target_by_feature(targets: Iterable, features: Iterable, split_value: Union[int, float]) -> Tuple[Iterable, Iterable]:
    sub_targets_1 = []
    sub_targets_2 = []
    
    features_list = list(features)
    targets_list = list(targets)
    
    for i, feature in enumerate(features_list):
        if feature <= split_value:
            sub_targets_1.append(targets_list[i])
        else: 
            sub_targets_2.append(targets_list[i])
            
    return (sub_targets_1, sub_targets_2)

print(split_target_by_feature([1,1,2,2,1,1,2,1,3,1,3,4,1], [3,1,9,9,1,5,2,1,7,1,3,4,9], 5))
        

([1, 1, 1, 1, 2, 1, 1, 3, 4], [2, 2, 3, 1])


---

### 3.2 Поиск оптимального "делителя"

Следующий шаг - поиск оптимального "делителя", который позволит получить таке две подгруппы основного списка `targets`, взвешенная оценка которых для указанной метрики будет минимальна.  
  
Используя предыдущий пример, оценим вариант двух разбиений:  
первый, когда "делитель" равен 5:

$$
T = [\color{red}{1}, \color{red}{1}, \color{blue}{2}, \color{blue}{2}, \color{red}{1}, \color{red}{1}, \color{red}{2}, \color{red}{1}, \color{blue}{3}, \color{red}{1}, \color{red}{3}, \color{red}{4}, \color{blue}{1}] \rightarrow [\color{red}{1,1,1,1,3,1,1,3,4}] \ , \ [\color{blue}{2,2,3,1}] \\
F = [\color{red}{3}, \color{red}{1}, \color{blue}{9}, \color{blue}{9}, \color{red}{1}, \color{red}{5}, \color{red}{2}, \color{red}{1}, \color{blue}{7}, \color{red}{1}, \color{red}{3}, \color{red}{4}, \color{blue}{9}] \rightarrow [\color{red}{3,1,1,5,2,1,1,3,4}] \ , \ [\color{blue}{9,9,7,9}]
$$

второй - когда "делитель" равен 3:

$$
T = [\color{red}{1}, \color{red}{1}, \color{blue}{2}, \color{blue}{2}, \color{red}{1}, \color{blue}{1}, \color{red}{2}, \color{red}{1}, \color{blue}{3}, \color{red}{1}, \color{red}{3}, \color{blue}{4}, \color{blue}{1}] \rightarrow [\color{red}{1,1,1,3,1,1,3}] \ , \ [\color{blue}{2,2,1,3,1,4}] \\
F = [\color{red}{3}, \color{red}{1}, \color{blue}{9}, \color{blue}{9}, \color{red}{1}, \color{blue}{5}, \color{red}{2}, \color{red}{1}, \color{blue}{7}, \color{red}{1}, \color{red}{3}, \color{blue}{4}, \color{blue}{9}] \rightarrow [\color{red}{3,1,1,2,1,1,3}] \ , \ [\color{blue}{9,9,5,7,9,4}]
$$

Допустим, что в качестве метрики выбрана этнропия.  
  
В таком случае, взвешенная оценка в первом варианте будет равна:

$$
Entropy_{total} = \frac{9}{13} * Entropy([1,1,1,1,2,1,1,3,4]) + \frac{4}{13} * Entropy([2,2,3,1]) = 0.694 + 0.320 = 1.014
$$

А во втором варианте:

$$
Entropy_{total} = \frac{7}{13} * Entropy([1,1,1,2,1,1,3]) + \frac{6}{13} * Entropy([2,2,1,3,1,4]) = 0.429 + 0.614 = 1.043
$$

Таким образом, `5` даёт меньшее значение взвешенной метрики по сравнению с `3`.

#### Задание: написать функцию, которая получает на вход функцию метрики `score_func`, основной список `targets`, вспомогательный список `features` и возвращает оптимальное значение "делителя":

In [58]:
def find_best_split_and_metric_value(score_func: Callable, targets: Iterable, features: Iterable) -> Tuple[Union[int, float], float]:
    targets_list = list(targets)
    features_list = list(features)
    
    unique_features = set(features) 
    
    for i, split_value in enumerate(unique_features):

        tuple_targets = split_target_by_feature(targets_list, features_list, split_value)
        weighted_metric = weighted_metric_score(score_func, tuple_targets[0], tuple_targets[1])

        if i == 0 or optimal_weighted_metric > weighted_metric:
            optimal_split_value = split_value
            optimal_weighted_metric = weighted_metric

    return (optimal_split_value, optimal_weighted_metric)
    

def find_best_split_value(score_func: Callable, targets: Iterable, features: Iterable) -> Union[int, float]:
    return find_best_split_and_metric_value(score_func, targets, features)[0]
    
    
print(find_best_split_value(count_entropy_score, [1,1,2,2,1,1,2,1,3,1,3,4,1], [3,1,9,9,1,5,2,1,7,1,3,4,9]))
print(find_best_split_value(count_entropy_score, [1,1,2,2,1,1,2,1,3,1,3,4,1], [3,3,2,2,2,2,2,2,7,1,3,4,9]))
print(find_best_split_value(count_entropy_score, [1,1,2,2,1,1,2,1,3,1,3,4,1], [0,1,2,3,0,2,3,1,2,2,3,0,0]))

1
2
1


---

### 3.3 Выбор оптимального вспомогательного списка

Расширим вводные данные: теперь помимо основного списка `targets`, нам даётся не один, а несколько вспомогательных списков `features` `F1`, `F2` ... `Fn`.

#### Задание: написать функцию, которая получает на вход функцию метрики `score_func`, основной список `targets`, группу вспомогательных списков`options` `F1`, `F2` ... `Fn` и возвращает индекс выбранного вспомогательного списка и оптимальное значение "делителя":

In [59]:
def find_best_split_value(score_func: Callable, targets: Iterable, options: List[Iterable]) -> Tuple[Iterable, Union[int, float]]:

    targets_list = list(targets)
    options_features_list = [list(option) for option in options]
    
    indexes = []
    
    for i, features in enumerate(options_features_list):
        
        split_metric_value = find_best_split_and_metric_value(score_func, targets, features)
        split_value = split_metric_value[0]
        weighted_metric = split_metric_value[1]
        
        if i == 0:
            indexes.append(i)
            optimal_split_value = split_value
            optimal_weighted_metric = weighted_metric
            
        elif optimal_weighted_metric > weighted_metric: 
            indexes = []
            indexes.append(i)
            optimal_split_value = split_value
            optimal_weighted_metric = weighted_metric
            
        elif optimal_weighted_metric == weighted_metric: 
            indexes.append(i)
            
    return (indexes, optimal_split_value)
    
    
print(find_best_split_value(count_entropy_score, [1,1,2,2,1,1,2,1,3,1,3,4,1],[[0,1,2,3,0,2,3,1,2,2,3,0,0],
                                                                              [3,3,2,2,2,2,2,2,7,1,3,4,9],
                                                                              [3,1,9,9,1,5,2,1,7,1,3,4,9],
                                                                              [0,1,2,3,0,2,3,1,2,2,3,0,0],
                                                                              [3,1,9,9,1,5,2,1,7,1,3,4,9]]))

([0, 3], 1)


---

# <center>Классы</center>

# 4. Узел (Node)

#### Задание: напишите класс `Node`
  
Он должен иметь следующие атрибуты:
- **id**       : int - **уникальный** номер для каждого экземпляра этого класса
- **values**   : Iterable - значения, хранимые в данном узле
- **parent**   : Optional `Node`- "родительский" узел, из которого был создан данный экземпляр (по умолчанию = None)
- **children** : Optional List `[Node]` - узлы, образованные из данного экземпляра (по умолчанию = None)  
  
Также он должен обладать следующими методами:
- **is_root** -> bool : проверяет, является ли данный узел "корневым" (т.е. имеется у него "родитель" или нет)
- **is_leaf** -> bool : проверяет, является ли данный узел терминальным (т.е. имеется у него "дети" или нет)

In [10]:
class NodeException(Exception):

    def __init__(self, message: str, node: Node):
        self.message = message
        self.node = node

    def __str__(self):
        return self.message + str(self.node)



class Node:
    def __init__(self, 
                 values: Iterable, 
                 parent: Optional["Node"] = None, 
                 children: Optional[List["Node"]] = None):

        self.uid = Node.generate_uid()
        self.values = values
        self.parent = parent
        self.children = children 
        
        if parent is not None:
            if parent.children is None:
                parent.children = []
            parent.children.append(self)
        
        if children is not None:
            self.add_children(children)
         
    def __str__(self):
        return f"\n uid: {self.uid}, values: {self.values}, parent: {self.parent if self.parent is None else self.parent.values}, children: {[item.values for item in self.children]}\n"
     
        
    @classmethod
    def generate_uid(cls):
        return random.randint(int(1e2), int(1e3) - 1)
    
    def is_root(self):
        return self.parent is None
    
    def is_leaf(self):
        return self.children is None

    def add_children(self, children: List["Node"]):
        for child in children:
            if child.parent is None:
                child.parent = self
                if self.children is None:
                    self.children = []
                self.children.append(child)
            elif child.parent != self:
                raise NodeException("Node child can have only one parent. Caused by ", self)
        return self
    
#         leaf_6 (exception)-> leaf_2               
# root -> leaf_1 -> leaf_2 -> leaf_4
#      -> leaf_3
#      -> leaf_5

try :
    leaf_1 = Node(values = [1])
    leaf_2 = Node(values = [2], parent = leaf_1)
    leaf_3 = Node(values = [3])

    root = Node(values = [0], children = [leaf_1, leaf_3])

    leaf_4 = Node(values = [4])
    leaf_2.add_children([leaf_4])

    leaf_5 = Node(values = [5], parent = root)
    print(root)
    
    leaf_6 = Node(values = [6], children = [leaf_2])
    
except NodeException as exception: 
    print(exception)


 uid: 855, values: [0], parent: None, children: [[1], [3], [1], [3], [5]]

Node child can have only one parent. Caused by 
 uid: 382, values: [6], parent: None, children: [[2]]



# 5. Условие (Condition) и Группа условий (Conditions Group)

#### Задание: напишите классы `Condition` и `ConditionsGroup`

**Condition** представляет собой описание следующего правила: `признак Х` - `меньше или равно | больше` - `значение`  
  
Таким образом, этот класс должен обладать соответствующими атрибутами:
- **feature** : str|int
- **greater** : bool
- **value**   : int|float

**ConditionsGroup** должен сохранять список разных условий и иметь атрибут **conditions**

<div class="alert alert-info">
    <center><strong>Задание на 10 баллов</strong></center>
</div>

Реализуйте операцию объединения экземпляров классов **Condition** и **ConditionsGroup** следующим образом:
- Если объединяются два экземпляра класса **Condition** и у обоих одинаковые атрибуты `feature` и `greater`, то объединение должно возвращать новый экземпляр **Condition** с тем значением атрибута `value`, который соответсвует логике

In [ ]:
c1 = Condition(feature="рост", greater=False, value=180)
c2 = Condition(feature="рост", greater=False, value=160)

c1 + c2 -> Condition(feature="рост", greater=False, value=160)


- В любом другом случае должен возвращаться новый экземпляр **ConditionsGroup**, и, если при этом между элементами этого экземпляра возможно произвести объединение первого типа - то это нужно сделать (таким образом, не будет нескольких элементов с одинаковыми `feature` и `greater`)

In [ ]:
c1 = Condition(feature="рост", greater=False, value=180)
c2 = Condition(feature="рост", greater=True, value=160)

cg_1 = c1 + c2
cg_1 -> ConditionsGroup(Condition(feature="рост", greater=False, value=180),
                        Condition(feature="рост", greater=True, value=160))

In [6]:
class Condition:
    
    def __init__(self, 
                 feature: Union[str, int], 
                 greater: bool, 
                 value: Union[int, float]):
        
        self.feature = feature
        self.greater = greater
        self.value = value
    
                 
    def __add__(self, condition: "Condition") -> Union["Condition", "ConditionsGroup"]:
        result = Condition.match_conditions(self, condition)
        if result is None:
            result = ConditionsGroup()
            result.add_condition(self)
            result.add_condition(condition)
        return result
    
    @classmethod 
    def match_conditions(cls, first_condition: "Condition", second_condition: "Condition") -> Optional["Condition"]:
        result = None
        if first_condition.feature == second_condition.feature and first_condition.greater == second_condition.greater:
            if first_condition.greater == second_condition.greater == True:
                result_value = first_condition.value if first_condition.value > second_condition.value else second_condition.value
            else:
                result_value = first_condition.value if first_condition.value <= second_condition.value else second_condition.value
            result = Condition(feature = first_condition.feature, greater = first_condition.greater, value = result_value)     
        return result
    
    
    def __str__(self):
        return f"\n feature: {self.feature}, greater: {self.greater}, value: {self.value}\n"

    
class ConditionsGroup:
    
    def __init__(self):
        self.conditions = []
        
    def __add__(self, input_var: Union[Condition, "ConditionsGroup"]) -> "ConditionsGroup":
        result = None
        if isinstance(input_var, Condition):
            result = ConditionsGroup.create_and_add_condition(self, input_var)
        elif isinstance(input_var, ConditionsGroup): 
            result = input_var
            if input_var.conditions:
                for condition in self.conditions:
                    new_group = ConditionsGroup.create_and_add_condition(result, condition)
                    result = new_group
        
        return result
             
    def __str__(self):
        return "ConditionsGroup:" + ' '.join(str(item) for item in self.conditions)
        
    def add_condition(self, new_condition: Condition) -> bool:
        is_matched = False
        
        if self.conditions:
            for condition in self.conditions:
                if Condition.match_conditions(condition, new_condition) is not None:
                    is_matched = True
                    
        if not is_matched:
            self.conditions.append(new_condition)
            
        return not is_matched
            
    @classmethod
    def create_and_add_condition(cls, group:"ConditionsGroup", new_condition: Condition) -> "ConditionsGroup":
        modified_group = ConditionsGroup()
        
        is_matched = False
        
        if group.conditions:
            for condition in group.conditions:
                match = Condition.match_conditions(condition, new_condition)
                if match is not None:
                    is_matched = True
                    modified_group.conditions.append(match)
                else: 
                    modified_group.conditions.append(condition)
                    
                    
        if not is_matched:
            modified_group.conditions.append(new_condition)
            
        return modified_group
        
ch1 = Condition(feature="рост", greater=False, value=180)
ch2 = Condition(feature="рост", greater=False, value=160)  
ch3 = Condition(feature="рост", greater=True, value=150)  

cw1 = Condition(feature="вес", greater=True, value=180)
cw2 = Condition(feature="вес", greater=False, value=160) 

ch4 = Condition(feature="рост", greater=True, value=200)

ch_sum1 = ch1 + ch2
ch_sum2 = ch_sum1 + ch3

cw_sum1 = cw1 + cw2
cw_sum2 = cw_sum1 + ch4

c_sum = cw_sum2 + ch_sum2

# print("ch1 + ch2 = ch_sum1 = ")
# print(ch_sum1)
print("ch_sum1 + ch3 = ch_sum2 = ")
print(ch_sum2)
# print("cw1 + cw2 = cw_sum1 = ")
# print(cw_sum1)
print("cw_sum1 + ch4 = cw_sum2 = ")
print(cw_sum2)

print("cw_sum2 + ch_sum2 = c_sum")
print(c_sum)
        
                 
                 

ch_sum1 + ch3 = ch_sum2 = 
ConditionsGroup:
 feature: рост, greater: False, value: 160
 
 feature: рост, greater: True, value: 150

cw_sum1 + ch4 = cw_sum2 = 
ConditionsGroup:
 feature: вес, greater: True, value: 180
 
 feature: вес, greater: False, value: 160
 
 feature: рост, greater: True, value: 200

cw_sum2 + ch_sum2 = c_sum
ConditionsGroup:
 feature: рост, greater: False, value: 160
 
 feature: рост, greater: True, value: 200
 
 feature: вес, greater: True, value: 180
 
 feature: вес, greater: False, value: 160



---

<center><h1>Удачи!</h1></center>